### PR TITLE
Use OpenAssistent/oasst1 dataset for custom dataset example

### PR DESCRIPTION
**Pull request by mreso**
# What does this PR do?

This PR replaces the samsum dataset in the custom dataset examples and uses the OpenAssistent/oasst1 dataset instead

Fixes # (issue)
N/A

## Feature/Issue validation/testing

Please describe the tests that you ran to verify your changes and relevant result summary. Provide instructions so it can be reproduced.
Please also list any relevant details for your test configuration.

- [X] pytest tests/
```
===================================================================================================== test session starts ======================================================================================================
platform linux -- Python 3.10.11, pytest-7.3.2, pluggy-1.0.0
rootdir: /home/ubuntu/llama-recipes
plugins: mock-3.11.1, anyio-3.7.0, hydra-core-1.0.7
collected 7 items

tests/test_finetuning.py ...                                                                                                                                                                                             [ 42%]
tests/test_train_utils.py .                                                                                                                                                                                              [ 57%]
tests/datasets/test_custom_dataset.py ..                                                                                                                                                                                 [ 85%]
tests/datasets/test_samsum_datasets.py .                                                                                                                                                                                 [100%]

======================================================================================================= warnings summary =======================================================================================================
src/llama_recipes/finetuning.py:5
  /home/ubuntu/llama-recipes/src/llama_recipes/finetuning.py:5: DeprecationWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html
    from pkg_resources import packaging

../miniconda3/envs/llama/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('mpl_toolkits')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../miniconda3/envs/llama/lib/python3.10/site-packages/pkg_resources/__init__.py:2871
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/pkg_resources/__init__.py:2871: DeprecationWarning: Deprecated call to `pkg_resources.declare_namespace('google')`.
  Implementing implicit namespace packages (as specified in PEP 420) is preferred to `pkg_resources.declare_namespace`. See https://setuptools.pypa.io/en/latest/references/keywords.html#keyword-namespace-packages
    declare_namespace(pkg)

../miniconda3/envs/llama/lib/python3.10/site-packages/bitsandbytes/cuda_setup/main.py:147
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/bitsandbytes/cuda_setup/main.py:147: UserWarning: /home/ubuntu/miniconda3/envs/llama did not contain ['libcudart.so', 'libcudart.so.11.0', 'libcudart.so.12.0'] as expected! Searching further paths...
    warn(msg)

../miniconda3/envs/llama/lib/python3.10/site-packages/torch/distributed/_shard/checkpoint/__init__.py:8
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/torch/distributed/_shard/checkpoint/__init__.py:8: DeprecationWarning: torch.distributed._shard.checkpoint will be deprecated, use torch.distributed.checkpoint instead
    warnings.warn(

tests/test_train_utils.py::test_gradient_accumulation
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/torch/cuda/memory.py:329: FutureWarning: torch.cuda.reset_max_memory_allocated now calls torch.cuda.reset_peak_memory_stats, which resets /all/ peak memory stats.
    warnings.warn(

tests/datasets/test_samsum_datasets.py::test_custom_dataset
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/dill/_dill.py:1705: PicklingWarning: Cannot locate reference to <class 'unittest.mock.MagicMock'>.
    warnings.warn('Cannot locate reference to %r.' % (obj,), PicklingWarning)

tests/datasets/test_samsum_datasets.py::test_custom_dataset
  /home/ubuntu/miniconda3/envs/llama/lib/python3.10/site-packages/dill/_dill.py:1707: PicklingWarning: Cannot pickle <class 'unittest.mock.MagicMock'>: unittest.mock.MagicMock has recursive self-references that trigger a RecursionError.
    warnings.warn('Cannot pickle %r: %s.%s has recursive self-references that trigger a RecursionError.' % (obj, obj.__module__, obj_name), PicklingWarning)

-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
================================================================================================ 7 passed, 8 warnings in 20.12s ================================================================================================
```

## Before submitting
- [X] Did you read the [contributor guideline](https://github.com/facebookresearch/llama-recipes/blob/main/CONTRIBUTING.md#pull-requests),
      Pull Request section?
- [X] Did you make sure to update the documentation with your changes?  
- [X] Did you write any new necessary tests?

### CODE REVIEW

File: tests/datasets/test_custom_dataset.py [hunk 1]
```diff
@@ -29,13 +28,15 @@ def test_custom_dataset(step_lr, optimizer, tokenizer, get_model, train, mocker)
...
+    tokenizer = args[3]
     
-    VAL_SAMPLES = 818
-    TRAIN_SAMPLES = 14732
-    CONCAT_SIZE = 2048
+    assert len(train_dataloader) == 226
+    assert len(eval_dataloader) == 2*226
+
+    STRING = tokenizer.decode(next(iter(train_dataloader))["input_ids"][0], skip_special_tokens=True)
+    EXPECTED_STRING = "[INST] Напиши функцию на языке swift, которая сортирует массив целых чисел, а затем выводит его на экран [/INST] Вот функция, "
```
Comment (chauhang) [inline tests/datasets/test_custom_dataset.py:37]: Will be better to use English example

File: tests/datasets/test_custom_dataset.py [hunk 2]
```diff
@@ -7,18 +7,17 @@
...
 @patch('llama_recipes.finetuning.train')
 @patch('llama_recipes.finetuning.LlamaForCausalLM.from_pretrained')
-@patch('llama_recipes.finetuning.LlamaTokenizer.from_pretrained')
 @patch('llama_recipes.finetuning.optim.AdamW')
 @patch('llama_recipes.finetuning.StepLR')
-def test_custom_dataset(step_lr, optimizer, tokenizer, get_model, train, mocker):
+def test_custom_dataset(step_lr, optimizer, get_model, train, mocker):
     from llama_recipes.finetuning import main
-        
-    tokenizer.return_value = mocker.MagicMock(side_effect=lambda x: {"input_ids":[len(x)*[0,]], "attention_mask": [len(x)*[0,]]})
```
Comment (chauhang) [inline tests/datasets/test_custom_dataset.py:16]: What is the reason for removing the mocked tokenizer?

File: tests/datasets/test_custom_dataset.py [hunk 3]
```diff
@@ -29,13 +28,15 @@ def test_custom_dataset(step_lr, optimizer, tokenizer, get_model, train, mocker)
     args, kwargs = train.call_args
     train_dataloader = args[1]
     eval_dataloader = args[2]
+    tokenizer = args[3]
```
Comment (chauhang) [inline tests/datasets/test_custom_dataset.py:31]: Please add steps for running this test. What are the expected values that user has to provide for the tokenizer and other params

**Review by chauhang — CHANGES_REQUESTED** [review]
@mreso Thanks for adding OpenAssistant dataset support. Please see the comments inline

File: tests/datasets/test_custom_dataset.py (hunk 1 above)
Comment (mreso) [reply tests/datasets/test_custom_dataset.py:37]: Sure, this was just the first example in the resulting dataset. Here we just want to see that the instruction tags are added correctly. But can select a different example as well.

File: tests/datasets/test_custom_dataset.py (hunk 3 above)
Comment (mreso) [reply tests/datasets/test_custom_dataset.py:31]: The test is self-contained and will run when a user calls the test suite with python -m tests. The args here just represent the argument that were given to the train function (which is actually mocked out) by the main fine-tuning routine which we're testing (besides the custom dataset preprocessing itself). We use the arguments given to mocked train function to check them against the expected values.

File: tests/datasets/test_custom_dataset.py (hunk 2 above)
Comment (mreso) [reply tests/datasets/test_custom_dataset.py:16]: The preprocessing calls the tokenizer so this part makes the test more realistic as we're actually preprocessing the data e2e (encoding + decoding).

**Review by chauhang — APPROVED** [review]

**Pull request merged**